### PR TITLE
py/objstringio: Prevent negative string position for BytesIO object.

### PR DESCRIPTION
**Pull request by tomlogic**
This is a change related to maintaining the current position in a BytesIO object.  Since MicroPython treats StringIO objects as identical to BytesIO, it affects those objects as well.

In existing code, we store a negative offset as a large unsigned value, which creates wrap-around issues in `stringio_write()`, causing it to miscalculate the need to call `vstr_add_len()` and potentially write bytes outside of the memory allocated to the vstr.

This change will increase code size, but attempts to follow behavior of CPython for negative offsets and repeated seeks past the current position until the offset exceeds some maximum.

Modifying test code to perform the write without checking the return of `seek()` will lead to a crash.

My original fix included changing `pos` of `mp_obj_stringio_t` from `mp_uint_t` to `mp_int_t`, but I don't think it's necessary.  `pos` should never exceed the maximum value of a signed `mp_int_t`.

### CODE REVIEW

**Comment by pfalcon** [comment]
You can't raise exceptions in these function, they're C-level, not Python-level.

I'm also not sure this code and explanation captures the problem well. First step should be figuring whether POSIX allows to seek before the start of file.

**Comment by tomlogic** [comment]
@pfalcon: Maybe I wasn't clear enough in my explanation.  My change doesn't have anything to do with the POSIX `seek()` function.  I'll update the description to add clarity.

This is code for a BytesIO (and StringIO) object, and I was referring to the `seek()` method on that object which is mapped to `stringio_ioctl()`, so it is possible to throw an exception.

I ran through different scenarios with a BytesIO object in CPython and can confirm that it does not allow seeking before the beginning of a file.  If you seek to byte 10, and then seek backwards 20 bytes from the current seek position, you'll be at byte 0 of the object.

**Comment by pfalcon** [comment]
@tomlogic : Perhaps I wasn't clear either: stream objects in MicroPython are modeled on POSIX files, and that would be the primary reference.

> Part of the problem

So, both here and in https://github.com/micropython/micropython/pull/3090 my understanding that the problem is wrap-around of address space. Or more specifically, the fact that addresses, represented as integers, wrap around, while MicroPython address space doesn't: we can't have an object start at address 0xffffffff and the continue to 0, 1, 2. But I don't see checks for that! You check different things, and at least in #3090 also introduce additional restrictions. That's why I wonder if you're interested to explore alternative solutions to the problem.

**Comment by tomlogic** [comment]
Let's look at this pull request as completely separate from #3090.

These changes are entirely constrained to a BytesIO/StringIO object for this pull request.  Being able to seek to before the start of object doesn't make sense and isn't supported by CPython.  Users are allowed to seek beyond the end, and even write to that space, and Python will insert `0x00` bytes to fill the gap.

You can't can't create a BytesIO/StringIO object that exceeds 2GB on an embedded system where your heap is measured in KB.  Therefore, there isn't a need to allow seeking to a position beyond the limit of a signed 32-bit int.

Yes, I'm definitely interested in exploring alternative solutions to the problem, and in getting feedback from people who know the entirety of the system better than I do.

**Comment by pfalcon** [comment]
> Let's look at this pull request as completely separate from #3090.

We can't, it's the same underlying problem, so solution should be consistent at least.

> Users are allowed to seek beyond the end, and even write to that space, and Python will insert 0x00 bytes to fill the gap.

Not Python, POSIX! That's the whole idea - we don't need to poke nose into CPython code on each occasion, we can pick up bad practices and licensing issues. There're more strong governing principles how stuff works.

> You can't can't create a BytesIO/StringIO object that exceeds 2GB on an embedded system where your heap is measured in KB.

In the other ticket, you mentioned 16-bit length. Using "compressed pointers" is pretty viable optimization for systems with heaps below or equal 64K, then it might be natural to "compress" sizes then you're talking about limiting user objects (bytes, bytearray, strings) to diminutive 32K, while system has 64K. Is it fair and good engineering?

**Comment by tomlogic** [comment]
Paul, I've tried to address your concerns by not adding a limit of signed integers versus unsigned.  I have left separate commits if there's a desire to compare my original attempt to this version which should allow for `BytesIO()` objects sized up to the maximum allowed by `mp_uint_t`.

I've updated `write()` to identify when a write would exceed that maximum, and to ensure that it always allocates more space when necessary.

I don't have a test platform where it's possible to have a `BytesIO()` object that grows close to that limit, so I do not have tests for that boundary.

I'm hoping we can at least agree on these changes before moving on to the more complicated #3090.

(And for the record, I have only looked at CPython code when working on the sequence fix.  I ran Python3 on Windows and Mac to explore the behavior of seeking before the beginning of a `BytesIO()` object, since we compare test code execution in Python3 as a way to verify correct MicroPython behavior.  And POSIX `lseek()` returns the new offset or -1 for error, which would imply that negative offsets are not possible.)

**Comment by pfalcon** [comment]
As I mentioned, I'd let @dpgeorge speak up next, how he sees it. Rest assured that your work on this is much appreciated and very important, but unlike e.g. https://github.com/micropython/micropython/pull/3091, the way to solve this is nor immediately obvious, so may need more consideration, rework, rinsing an repeating.  

If you really want to continue doing code changes, please pay attention to previous review comments:

> You can't raise exceptions in these function, they're C-level, not Python-level.

Stream interface functions are supposed to return error in *errptr, not raise exceptions. Grep for existing implementations.

> And POSIX lseek() returns the new offset or -1 for error, which would imply that negative offsets are not possible.

That's not true: http://pubs.opengroup.org/onlinepubs/009695399/functions/lseek.html

> Upon successful completion, the resulting offset, as measured in bytes from the beginning of the file, shall be returned. Otherwise, (off_t)-1 shall be returned, errno shall be set to indicate the error, and the file offset shall remain unchanged.

As experience developer, you immediately will make an implication that off_t can be an unsigned type, and only its "all ones" value has a special meaning, all other values are valid offsets "from the beginning of the file", which cannot be negative.

**Comment by tomlogic** [comment]
> Stream interface functions are supposed to return error in *errptr, not raise exceptions.

OK, I have updated the `ioctl()` and `write()` calls accordingly.  I didn't understand your previous statement at first because `write()` calls `vstr_add_len()` which will throw a `MemoryError` if it can't allocate the memory.  Throwing exceptions *can* happen, but I should be returning an error.  I get it now, and remember using this interface when writing `read()` and `write()` interfaces for our sockets.

> That's not true...

And you've lost me here.  It sounds like you're agreeing with my statement _negative offsets are not possible_ when you say _all other values are valid offsets "from the beginning of the file", which cannot be negative._  My argument was that seeking to position -1 (return -1) would look identical to an error condition (set errno and return -1) if you allowed negative positions.

And further referencing the `lseek()` documentation fails to provide a simple answer to negative file positions:

> An invalid file offset that would cause [EINVAL] to be returned may be both implementation-defined and device-dependent (for example, memory may have few invalid values). A negative file offset may be valid for some devices in some implementations.

> The POSIX.1-1990 standard did not specifically prohibit lseek() from returning a negative offset. Therefore, an application was required to clear errno prior to the call and check errno upon return to determine whether a return value of ( off_t)-1 is a negative offset or an indication of an error condition. The standard developers did not wish to require this action on the part of a conforming application, and chose to require that errno be set to [EINVAL] when the resulting file offset would be negative for a regular file, block special file, or directory.

In the current `BytesIO()` case, I hope that following the CPython behavior of setting the file position to 0 is acceptable.  We could go the `EINVAL` route, but this technically isn't a "regular file, block special file, or directory".

Leaving my tortured commit history in place for now, and will squash down to two commits -- one for the updated test and one for the changes to `objstringio.c`.

**Comment by pfalcon** [comment]
> OK, I have updated the ioctl() and write() calls accordingly. I didn't understand your previous statement at first because write() calls vstr_add_len() which will throw a MemoryError if it can't allocate the memory.

That of course would mean that cases of throwing exceptions should be fixed, not that we should start to throw exceptions freely. (In this case, it's low priority, the idea of not throwing exceptions is to allow C code to recover from error easily, and there can be only so much recover from lack of memory.)

> I hope that following the CPython behavior

Once again, CPython should NOT be treated as the ultimate authority. There can be only 2 cases:

1. CPython has bugs and idiosyncrasies. 
2. CPython follows "more ultimate" authorities in its behavior, e.g. POSIX.

In both cases, recourse to CPython behavior optional. Here's a simple testcase which should have been posted as a seed for the discussion:

~~~
#include <sys/types.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>

int main()
{
    int fd = open("lseek.c", O_RDONLY);
    printf("lseek: %lx\n", lseek(fd, -1, SEEK_CUR));
    printf("errno: %d\n", errno);
}
~~~

~~~
$ ./lseek 
lseek: ffffffffffffffff
errno: 22
~~~

**Comment by pfalcon** [comment]
Updated testcase:
~~~
#include <sys/types.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>

int main()
{
    int fd = open("lseek.c", O_RDONLY);
    printf("lseek: %lx\n", lseek(fd, -1, SEEK_CUR));
    printf("errno: %d\n", errno);

    lseek(fd, 1, SEEK_SET);
    printf("lseek: %lx\n", lseek(fd, -2, SEEK_CUR));
    printf("errno: %d\n", errno);
}
~~~

~~~
lseek: ffffffffffffffff
errno: 22
lseek: ffffffffffffffff
errno: 22
~~~

File: py/objstringio.c [hunk 1]
```diff
@@ -68,23 +68,30 @@ STATIC mp_uint_t stringio_read(mp_obj_t o_in, void *buf, mp_uint_t size, int *er
 STATIC mp_uint_t stringio_write(mp_obj_t o_in, const void *buf, mp_uint_t size, int *errcode) {
     (void)errcode;
     mp_obj_stringio_t *o = MP_OBJ_TO_PTR(o_in);
+    mp_uint_t new_pos, org_len;
```
Comment (pfalcon) [inline py/objstringio.c]: Please declare vars whey they're used.

File: py/objstringio.c [hunk 2]
```diff
@@ -68,23 +68,30 @@ STATIC mp_uint_t stringio_read(mp_obj_t o_in, void *buf, mp_uint_t size, int *er
...
+    
     check_stringio_is_open(o);
-    mp_int_t remaining = o->vstr->alloc - o->pos;
-    mp_uint_t org_len = o->vstr->len;
-    if ((mp_int_t)size > remaining) {
+    new_pos = o->pos + size;
```
Comment (pfalcon) [inline py/objstringio.c]: Should be "mp_uint_t new_pos".

File: py/objstringio.c [hunk 3]
```diff
@@ -68,23 +68,30 @@ STATIC mp_uint_t stringio_read(mp_obj_t o_in, void *buf, mp_uint_t size, int *er
...
     if (o->pos > org_len) {
         memset(o->vstr->buf + org_len, 0, o->pos - org_len);
     }
     memcpy(o->vstr->buf + o->pos, buf, size);
-    o->pos += size;
-    if (o->pos > o->vstr->len) {
-        o->vstr->len = o->pos;
+    o->pos = new_pos;
+    if (new_pos > o->vstr->len) {
+        o->vstr->len = new_pos;
```
Comment (pfalcon) [inline py/objstringio.c]: Changes up and including this line appear to be correct, please resubmit as a separate PR with a single commit. Thanks.

**Comment by tomlogic** [comment]
I'll create a new pull request with just the fixes to `stringio_write()`.  Should I prepare separate pull requests for two different behaviors when seeking to before the start of the file?

1) Reset position to 0.
2) Return an EINVAL error.

Or have you decided that you just want to follow the POSIX `lseek()` behavior (EINVAL)?

**Comment by pfalcon** [comment]
> I'll create a new pull request with just the fixes to stringio_write().

Thanks!

As for the rest, let's handle it one by one (I'm pretty busy now and don't have free cycles to give it enough thought, we may need to wait until after 1.9 release. Thanks for understanding.)

**Pull request closed without merging**